File: dino_game.py
import os
import sys
import pygame
import random
import time

def resource_path(relative_path):
    """ Get absolute path to resource, works for dev and for PyInstaller """
    try:
        # PyInstaller creates a temp folder and stores path in _MEIPASS
        base_path = sys._MEIPASS
    except Exception:
        base_path = os.path.abspath(".")

    return os.path.join(base_path, relative_path)

# Pygame 초기화
pygame.init()

# 화면 크기 설정
screen_width = 1000
screen_height = 800
screen = pygame.display.set_mode((screen_width, screen_height))
pygame.display.set_caption("병역기피게임")

# 색상 설정
white = (255, 255, 255)
red = (255, 0, 0)

# 장애물과 공룡 크기 설정
obstacle_width = 100  # 두 배로 설정
obstacle_height = 100  # 두 배로 설정
dino_width = int(obstacle_width * 0.8)  # 현재의 0.8배로 설정
dino_height = int(obstacle_height * 0.8)  # 현재의 0.8배로 설정
dino_speed = 10

# 공룡 이미지 로드 및 크기 조정
dino_image_normal = pygame.image.load(resource_path('dino.png'))
dino_image_normal = pygame.transform.scale(dino_image_normal, (dino_width, dino_height))

dino_image_broken = pygame.image.load(resource_path('dino_broken.png'))
dino_image_broken = pygame.transform.scale(dino_image_broken, (dino_width, dino_height))

# 장애물 이미지 로드 및 크기 조정
obstacle_image1 = pygame.image.load(resource_path('obstacle1.png'))
obstacle_image1 = pygame.transform.scale(obstacle_image1, (obstacle_width, obstacle_height))

obstacle_image2 = pygame.image.load(resource_path('obstacle2.png'))
obstacle_image2 = pygame.transform.scale(obstacle_image2, (obstacle_width, obstacle_height))

obstacle_images = [obstacle_image1, obstacle_image2]

# 배경 이미지 로드
background_image = pygame.image.load(resource_path('background.png'))
background_image = pygame.transform.scale(background_image, (screen_width, screen_height))

# 게임 안내문 이미지 로드
instructions_image = pygame.image.load(resource_path('instructions.png'))
instructions_image = pygame.transform.scale(instructions_image, (screen_width, screen_height))

# 장애물 속도 설정
obstacle_speed = 20  # 장애물 속도를 두 배로 설정

# 게임 상태 변수
running = True

# 최근 3번의 스코어를 저장할 리스트
recent_scores = []

# 시계 객체 초기화
clock = pygame.time.Clock()

# 게임 초기화 함수
def init_game():
    global dino_x, dino_y, dino_image, game_over, obstacles, obstacle_timer, start_time, score, total_score, blink_timer, show_instructions
    dino_x = screen_width // 2
    dino_y = screen_height - dino_height - 10
    dino_image = dino_image_normal
    game_over = False
    obstacles = []
    obstacle_timer = 0
    start_time = time.time()
    score = 0
    total_score = 0
    blink_timer = 0
    show_instructions = True

# 새로운 장애물을 생성하는 함수
def create_obstacle():
    for _ in range(4):  # 한 번에 4개의 장애물을 생성
        obstacle_x = random.randint(0, screen_width - obstacle_width)
        obstacle_y = -obstacle_height
        obstacle_img = random.choice(obstacle_images)
        obstacles.append([obstacle_x, obstacle_y, obstacle_img])

# 장애물을 업데이트하는 함수
def update_obstacles():
    for obstacle in obstacles:
        obstacle[1] += obstacle_speed
    if len(obstacles) > 0 and obstacles[0][1] > screen_height:
        obstacles.pop(0)

# 게임 실행 함수
def run_game():
    global dino_x, dino_y, running, game_over, dino_image, clock, obstacle_timer, score, total_score, blink_timer, show_instructions, recent_scores

    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
            elif event.type == pygame.KEYDOWN:
                if show_instructions:
                    show_instructions = False  # 게임 안내문을 닫고 게임 시작
                    start_time = time.time()  # 게임 시작 시간을 재설정
                elif event.key == pygame.K_r and game_over:
                    recent_scores.append(total_score)  # 최근 스코어에 추가
                    if len(recent_scores) > 3:
                        recent_scores.pop(0)  # 최근 3개 스코어만 유지
                    init_game()

        keys = pygame.key.get_pressed()
        current_speed = dino_speed
        if keys[pygame.K_SPACE]:
            current_speed = dino_speed * 2

        if not show_instructions and not game_over:
            if keys[pygame.K_LEFT] and dino_x > 0:
                dino_x -= current_speed
            if keys[pygame.K_RIGHT] and dino_x < screen_width - dino_width:
                dino_x += current_speed
            if keys[pygame.K_UP] and dino_y > 0:
                dino_y -= current_speed
            if keys[pygame.K_DOWN] and dino_y < screen_height - dino_height:
                dino_y += current_speed

            obstacle_timer -= 1
            if obstacle_timer <= 0:
                create_obstacle()
                obstacle_timer = random.randint(10, 30)  # 장애물이 떨어지는 시간차를 조절

            update_obstacles()

            # 충돌 감지
            for obstacle in obstacles:
                if (dino_x < obstacle[0] + obstacle_width and
                    dino_x + dino_width > obstacle[0] and
                    dino_y < obstacle[1] + obstacle_height and
                    dino_y + dino_height > obstacle[1]):
                    game_over = True
                    dino_image = dino_image_broken
                    total_score = score  # 게임 종료 시 누적 점수 설정
                    recent_scores.append(total_score)  # 최근 스코어에 추가
                    if len(recent_scores) > 3:
                        recent_scores.pop(0)  # 최근 3개 스코어만 유지

            # 점수 계산
            score = int((time.time() - start_time) * 10)

        screen.blit(background_image, (0, 0))  # 배경 이미지 그리기

        if show_instructions:
            screen.blit(instructions_image, (0, 0))  # 게임 안내문 이미지 그리기
        elif game_over:
            # 큰 폰트 설정
            big_font = pygame.font.SysFont(None, 110, bold=True)
            font = pygame.font.SysFont(None, 55)
            blink_timer += 1
            if blink_timer % 30 < 15:  # 15프레임마다 깜빡임
                text1 = big_font.render("YOU'RE DEAD!!", True, red)
                screen.blit(text1, [screen_width // 6, screen_height // 3 - 50])
            text2 = font.render("Press 'R' to Restart", True, red)
            screen.blit(text2, [screen_width // 6, screen_height // 3 + 50])
            # 누적 점수 표시
            score_text = big_font.render(f"${total_score}", True, (0, 0, 0))
            screen.blit(score_text, [screen_width - 200, 20])
            # 최근 스코어 표시
            recent_scores_text = font.render("Recent Scores:", True, red)
            screen.blit(recent_scores_text, [screen_width // 6, screen_height // 2])
            for i, recent_score in enumerate(recent_scores):
                score_display = font.render(f"{i + 1}. ${recent_score}", True, red)
                screen.blit(score_display, [screen_width // 6, screen_height // 2 + 40 * (i + 1)])
        else:
            # 실시간 점수 표시
            score_font = pygame.font.SysFont(None, 110, bold=True)
            score_text = score_font.render(f"${score}", True, (0, 0, 0))
            screen.blit(score_text, [screen_width - 200, 20])

            screen.blit(dino_image, (dino_x, dino_y))

            for obstacle in obstacles:
                screen.blit(obstacle[2], (obstacle[0], obstacle[1]))  # 장애물 이미지 그리기

        pygame.display.update()
        clock.tick(30)

# 게임 초기화
init_game()

# 게임 루프 실행
run_game()

# Pygame 종료
pygame.quit()

File: dino_game.js
const canvas = document.getElementById('gameCanvas');
const ctx = canvas.getContext('2d');

// 게임 설정
const screenWidth = canvas.width;
const screenHeight = canvas.height;
const dinoWidth = 80;
const dinoHeight = 80;
const dinoSpeed = 10;
const obstacleWidth = 100;
const obstacleHeight = 100; // 장애물 높이를 원래대로 유지
const obstacleSpeed = 5 * 1.2; // 장애물 속도를 1.2배로 증가
let obstacleTimer = 0;

let dinoX = screenWidth / 2;
let dinoY = screenHeight - dinoHeight - 10;
let obstacles = [];
let gameOver = false;
let score = 0;
let startTime = Date.now();
let showInstructions = true;
let recentScores = [];
let blinkTimer = 0;
let isAccelerating = false; // 스페이스바로 가속하는 기능 추가
let specialImageDisplayed = false; // 특정 이미지를 표시했는지 여부

// 깜빡임 설정
let blinkStartTime = Date.now();
let isBlinking = false;

const BLINK_INTERVAL = 10000; // 10초 간격
const BLINK_DURATION = 3000; // 3초 동안 깜빡임

// 이미지 로드
const dinoImageNormal = new Image();
dinoImageNormal.src = 'dino.png';
const dinoImageBroken = new Image();
dinoImageBroken.src = 'dino_broken.png';
const obstacleImage1 = new Image();
obstacleImage1.src = 'obstacle1.png';
const obstacleImage2 = new Image();
obstacleImage2.src = 'obstacle2.png';
const backgroundImage = new Image();
backgroundImage.src = 'background.png';
const instructionsImage = new Image();
instructionsImage.src = 'instructions.png';
const specialImage = new Image();
specialImage.src = 'special.png'; // 특정 이미지 파일 로드

function createObstacle() {
    for (let i = 0; i < 8; i++) { // 장애물 수를 2배로 증가 (원래는 4개였음)
        let obstacleX = Math.random() * (screenWidth - obstacleWidth);
        let obstacleY = -obstacleHeight - (i * 200); // 장애물들이 서로 겹치지 않게 y 위치를 조정
        let obstacleImg = Math.random() > 0.5 ? obstacleImage1 : obstacleImage2;
        obstacles.push({ x: obstacleX, y: obstacleY, img: obstacleImg });
    }
}

function updateObstacles() {
    for (let obstacle of obstacles) {
        obstacle.y += obstacleSpeed;
    }
    obstacles = obstacles.filter(obstacle => obstacle.y < screenHeight); // 화면을 벗어난 장애물 제거
}

function detectCollision() {
    if (isBlinking) return; // 깜빡이는 동안 충돌 감지 무시
    for (let obstacle of obstacles) {
        if (dinoX < obstacle.x + obstacleWidth &&
            dinoX + dinoWidth > obstacle.x &&
            dinoY < obstacle.y + obstacleHeight &&
            dinoY + dinoHeight > obstacle.y) {
            gameOver = true;
        }
    }
}

function updateScore() {
    score = Math.floor((Date.now() - startTime) / 1000) * 10;
}

function draw() {
    ctx.clearRect(0, 0, screenWidth, screenHeight); // 캔버스를 깨끗하게 지움
    ctx.drawImage(backgroundImage, 0, 0, screenWidth, screenHeight);

    if (showInstructions) {
        ctx.drawImage(instructionsImage, 0, 0, screenWidth, screenHeight);
    } else if (specialImageDisplayed) {
        ctx.drawImage(specialImage, 0, 0, screenWidth, screenHeight); // 특정 이미지를 표시
    } else if (gameOver) {
        ctx.font = '110px Arial';
        ctx.fillStyle = 'red';
        blinkTimer++;
        if (blinkTimer % 30 < 15) {
            ctx.font = 'bold 110px Arial'; // "YOU'RE DEAD!!"의 글자 굵기를 두 배로 설정
            ctx.fillText("YOU'RE DEAD!!", screenWidth / 6 - 60, screenHeight / 3 - 50); // 왼쪽으로 3cm 이동
        }
        ctx.font = '55px Arial';
        ctx.fillText("Press 'R' to Restart", screenWidth / 6, screenHeight / 3 + 50);
        ctx.fillStyle = 'black';
        ctx.fillText(`$${score}`, screenWidth - 200 + 20, 20 + 45); // 오른쪽으로 2cm, 아래로 3cm 이동
        ctx.fillStyle = 'red';
        ctx.fillText("Recent Scores:", screenWidth / 6, screenHeight / 2);

        // 현재 점수 포함하여 최근 3회 기록 표시
        const displayScores = [...recentScores, score].slice(-3); 
        for (let i = 0; i < displayScores.length; i++) {
            ctx.fillText(`${i + 1}. $${displayScores[i]}`, screenWidth / 6, screenHeight / 2 + 40 + (i * 40)); // 1cm씩 간격 띄움
        }
    } else {
        ctx.font = '55px Arial'; // 게임 중의 글자 크기를 게임 오버 화면과 일치시킴
        ctx.fillStyle = 'black';
        ctx.fillText(`$${score}`, screenWidth - 200 + 20, 20 + 45); // 오른쪽으로 2cm, 아래로 3cm 이동

        if (isBlinking) {
            if (Math.floor(Date.now() / 200) % 2) { // 0.2초 간격으로 깜빡임
                ctx.globalAlpha = 0.5; // 반투명 효과
            } else {
                ctx.globalAlpha = 1.0;
            }
        } else {
            ctx.globalAlpha = 1.0;
        }

        ctx.drawImage(dinoImageNormal, dinoX, dinoY, dinoWidth, dinoHeight);
        ctx.globalAlpha = 1.0; // 다시 불투명으로 설정

        for (let obstacle of obstacles) {
            ctx.drawImage(obstacle.img, obstacle.x, obstacle.y, obstacleWidth, obstacleHeight);
        }
    }
}

function gameLoop() {
    if (!showInstructions && !gameOver) {
        if (obstacleTimer <= 0) {
            createObstacle();
            obstacleTimer = 150; // 장애물 생성 주기
        }
        obstacleTimer--;
        updateObstacles();
        detectCollision();
        updateScore();

        // 특정 점수 이상 시 specialImage 표시
        if (score > 1100 && !specialImageDisplayed) {
            specialImageDisplayed = true;
        }

        // 깜빡임 상태 관리
        if (!isBlinking && Date.now() - blinkStartTime >= BLINK_INTERVAL) {
            isBlinking = true;
            blinkStartTime = Date.now();
        }
        if (isBlinking && Date.now() - blinkStartTime >= BLINK_DURATION) {
            isBlinking = false;
            blinkStartTime = Date.now();
        }
    }
    draw();
    requestAnimationFrame(gameLoop);
}

window.addEventListener('keydown', (e) => {
    if (showInstructions) {
        showInstructions = false;
        startTime = Date.now();
    } else if (gameOver && e.key === 'r') {
        recentScores.push(score);
        if (recentScores.length > 2) {
            recentScores.shift();
        }
        dinoX = screenWidth / 2;
        dinoY = screenHeight - dinoHeight - 10;
        obstacles = [];
        gameOver = false;
        score = 0;
        specialImageDisplayed = false; // 게임 재시작 시 specialImage 초기화
        blinkStartTime = Date.now(); // 깜빡임 타이머 초기화
        startTime = Date.now();
    } else {
        switch (e.key) {
            case 'ArrowLeft':
                dinoX = Math.max(dinoX - (isAccelerating ? dinoSpeed * 2 : dinoSpeed), 0);
                break;
            case 'ArrowRight':
                dinoX = Math.min(dinoX + (isAccelerating ? dinoSpeed * 2 : dinoSpeed), screenWidth - dinoWidth);
                break;
            case 'ArrowUp':
                dinoY = Math.max(dinoY - (isAccelerating ? dinoSpeed * 2 : dinoSpeed), 0);
                break;
            case 'ArrowDown':
                dinoY = Math.min(dinoY + (isAccelerating ? dinoSpeed * 2 : dinoSpeed), screenHeight - dinoHeight);
                break;
            case ' ':
                isAccelerating = true; // 스페이스바로 가속 시작
                break;
        }
    }
});

window.addEventListener('keyup', (e) => {
    if (e.key === ' ') {
        isAccelerating = false; // 스페이스바로 가속 종료
    }
});

gameLoop();
